lib/concurrent.js

```diff
@@ -10,86 +12,85 @@ function Concurrent(tests, bail) {
 		throw new TypeError('Class constructor Concurrent cannot be invoked without \'new\'');
 	}
 
+	if (!tests) {
+		throw new Error('Concurrent items can\'t be undefined');
+	}
+
+	this.results = [];
+	this.passed = true;
+	this.reason = null;
 	this.tests = tests;
-	this.bail = bail;
+	this.bail = bail || false;
+
+	Object.keys(Concurrent.prototype).forEach(function (key) {
+		this[key] = this[key].bind(this);
+	}, this);
 }
 
 Concurrent.prototype.run = function () {
-	var bail = this.bail;
-	var tests = this.tests;
-	var sync = true;
-	var passed = true;
-	var reason;
-
-	var results = [];
-
-	function addAsync(result) {
-		if (!result.passed) {
-			if (passed) {
-				passed = false;
-				reason = result.reason;
-
-				if (bail) {
-					throw BAIL_ERROR;
-				}
-			}
+	var results;
+
+	try {
+		results = this.tests.map(this._runTest);
+	} catch (err) {
+		if (err instanceof AvaError) {
+			return this._results();
 		}
 
-		return result;
+		throw err;
 	}
 
-	for (var i = 0; i < tests.length; i++) {
-		var result = tests[i].run();
-
-		if (isPromise(result)) {
-			sync = false;
-			results.push(result.then(addAsync));
-		} else {
-			results.push(result);
-
-			if (!result.passed) {
-				if (bail) {
-					return {
-						passed: false,
-						result: results,
-						reason: result.reason
-					};
-				} else if (passed) {
-					passed = false;
-					reason = result.reason;
-				}
-			}
-		}
-	}
+	var isAsync = results.some(isPromise);
 
-	if (sync) {
-		return {
-			passed: passed,
-			reason: reason,
-			result: results
-		};
+	if (isAsync) {
+		return Promise.all(results)
+			.catch(AvaError, noop)
+			.then(this._results);
 	}
 
-	var ret = Promise.all(results).then(function (results) {
-		return {
-			passed: passed,
-			reason: reason,
-			result: results
-		};
-	});
-
-	if (bail) {
-		ret = ret.catch(function (err) {
-			if (err !== BAIL_ERROR) {
-				throw err;
-			}
-
-			return {
-				passed: false,
-				reason: reason
-			};
+	return this._results();
+};
+
+Concurrent.prototype._runTest = function (test, index) {
+	var result = test.run();
+
+	if (isPromise(result)) {
+		var self = this;
+
+		return result.then(function (result) {
+			return self._addResult(result, index);
 		});
 	}
 
-	return ret;
+	return this._addResult(result, index);
+};
+
+Concurrent.prototype._addResult = function (result, index) {
+	// always save result when not in bail mode or all previous tests pass
+	if ((this.bail && this.passed) || !this.bail) {
+		this.results[index] = result;
+	}
+
+	if (result.passed === false) {
+		this.passed = false;
+
+		// only set reason once
+		if (!this.reason) {
+			this.reason = result.reason;
+		}
+
+		if (this.bail) {
+			throw new AvaError('Error in Concurrent while in bail mode');
```

Comment:
This error will never be shown to a user, it's just for us. Did not want to leave AvaError without a message. It's for us, like a code comment, to indicate that this error causes execution to bail.